### PR TITLE
Touch support for desktop platforms

### PR DESCRIPTION
**Pull request by Susko3**
- Closes https://github.com/ppy/osu-framework/issues/1514
- Closes https://github.com/ppy/osu-framework/issues/5033
- Closes https://github.com/ppy/osu/issues/18889

Touch currently uses [mouse emulation](https://github.com/libsdl-org/SDL/blob/6a2e6c82a0764a00123447d93999ebe14d509aa8/include/SDL_hints.h#L1367-L1376) on desktop platforms. This PR sets the correct flag and implements [`SDL_TouchFingerEvent`](https://wiki.libsdl.org/SDL_TouchFingerEvent) handling. Windows also emulates mouse events for touch, and those events have also been properly ignored in `WindowsMouseHandler`.

Initial testing [by peppy on discord](https://discord.com/channels/188630481301012481/589331078574112768/993813779160760340).

Raw input touch event detection [inspired](https://github.com/libsdl-org/SDL/blob/85bbf8eec9f5e422447807b6cabec8c9ec0859d9/src/video/windows/SDL_windowsevents.c#L825) [by](https://github.com/libsdl-org/SDL/blob/85bbf8eec9f5e422447807b6cabec8c9ec0859d9/src/video/windows/SDL_windowsevents.c#L521-L537) [SDL](https://github.com/libsdl-org/SDL/blob/85bbf8eec9f5e422447807b6cabec8c9ec0859d9/src/video/windows/SDL_windowsevents.c#L509-L511) and confirmed by [Microsoft Docs](https://docs.microsoft.com/en-us/windows/win32/tablet/system-events-and-mouse-messages). (The `0x82` check seems to be unnecessary, as it's already covered by `0x80`.)

`getTouchSource` / `assignNextAvailableTouchSource` logic taken from iOS touch handler.

### CODE REVIEW

**Comment by peppy** [comment]
All working well now, until raw input is turned on, where things still seem to fall apart (movement seems to work but drag operations don't work, and tap operations only sometimes work).

I'll investigate further tomorrow if the issue isn't something obvious.

File: osu.Framework/Platform/Windows/WindowsMouseHandler.cs [hunk 1]
```diff
@@ -64,6 +64,10 @@ private IntPtr onWndProc(IntPtr userData, IntPtr hWnd, uint message, ulong wPara
             if (message != Native.Input.WM_INPUT)
                 return IntPtr.Zero;
 
+            if (Native.Input.IsTouchEvent(Native.Input.GetMessageExtraInfo()))
+                // touch events are handled by TouchHandler
+                return IntPtr.Zero;
```
Comment (peppy) [inline osu.Framework/Platform/Windows/WindowsMouseHandler.cs:69]: So this never gets hit in testing.

I also tested

```csharp
if (Native.Input.IsTouchEvent(data.Mouse.ExtraInformation))
```

and it also returns false even for touch.

Using `mouse.ExtraInformation>0` (see incorrect commented out version below) *does* work.

For reference, `ExtraInformation` is `139` in the touch event I tested.

Comment (Susko3) [reply osu.Framework/Platform/Windows/WindowsMouseHandler.cs:69]: What if you comment out our [custom handling](https://github.com/ppy/osu-framework/blob/ba1385330cc501f34937e08257e586c84e35d772/osu.Framework/Platform/Windows/WindowsGameHost.cs#L62-L68) and let SDL do everything? Curious if SDL will properly filter out the raw mouse touch events.

Maybe the `|| Native.Input.GetMessageExtraInfo() & 0x82 == 0x82` check really is necessary.

Comment (frenzibyte) [reply osu.Framework/Platform/Windows/WindowsMouseHandler.cs:69]: Upon blaming through SDL, they seem to have been using the same `IsTouchEvent` method used here, until 2 years ago, in which they added an additional condition [with the intent of fixing touch input being sent as regular mouse messages](https://github.com/libsdl-org/SDL/commit/0721931f5b92c96ff19bf936186bc4cf3dac38a1), I've also found this [thread](https://answers.microsoft.com/en-us/windows/forum/all/getmessageextrainfo-signature-missing-for-touch/f7ad7289-fd15-4061-98c5-2d90a176fe7c) from a year ago as well, which touches on the fact that the touch signature (`0xFF515700`) being missing (along with the touch identifier bit, except that it does seem to be present with `ExtraInformation` here).

I think that concludes the necessity of the `GetMessageExtraInfo() & 0x82 == 0x82` condition.

Comment (Susko3) [reply osu.Framework/Platform/Windows/WindowsMouseHandler.cs:69]: I can get behind only checking for `0x80` (and omitting the `MI_WP_SIGNATURE_MASK` check), but the `0x02` there confuses me.

> https://docs.microsoft.com/en-us/windows/win32/tablet/system-events-and-mouse-messages:
> The lower 8 bits returned from `GetMessageExtraInfo` are variable. Of those bits, 7 (the lower 7, masked by 0x7F) are used to represent the cursor ID, zero for the mouse or a variable value for the pen ID.

Checking for `0x02` is probably wrong. [This thread](https://docs.microsoft.com/en-us/answers/questions/285548/getmessageextrainfo-signature-missing-for-touch-in.html) confirms that; the received values were `0xC7/C8/F7/F8/F9` -- some of which don't have `0x2` set.

Why I think just `0x80` is fine:
> https://docs.microsoft.com/en-us/windows/win32/tablet/system-events-and-mouse-messages (cont.):
> Additionally, in Windows Vista, the eighth bit, masked by 0x80, is used to differentiate touch input from pen input (0 = pen, 1 = touch).

---

> https://discord.com/channels/188630481301012481/589331078574112768994938919701467226
> `GetMessageExtraInfo` always seemed to return 0

It seems `GetMessageExtraInfo` is not reliable, so `mouse.ExtraInformation` is what we want to use. It contains at least some useful information, as peppy said [above](https://github.com/ppy/osu-framework/pull/5299#discussion_r916723692).

**Comment by peppy** [comment]
Works perfectly now!

File: osu.Framework/Platform/Windows/WindowsMouseHandler.cs [hunk 2]
```diff
@@ -64,6 +64,11 @@ private IntPtr onWndProc(IntPtr userData, IntPtr hWnd, uint message, ulong wPara
             if (message != Native.Input.WM_INPUT)
                 return IntPtr.Zero;
 
+            if (Native.Input.IsTouchEvent(Native.Input.GetMessageExtraInfo()))
+                // sometimes GetMessageExtraInfo returns 0, so additionally, mouse.ExtraInformation is checked below.
+                // touch events are handled by TouchHandler
+                return IntPtr.Zero;
```
Comment (peppy) [inline osu.Framework/Platform/Windows/WindowsMouseHandler.cs:70]: I'd probably argue this is unnecessary at this point, but probably no harm in leaving it here..

**Review by peppy — APPROVED** [review]

**Pull request merged**